Lib/test/test_builtin.py

```diff
@@ -1686,44 +1676,94 @@ def run_child(self, child, terminal_input):
 
         return lines
 
+    def final_output(self, fd):
+        # Beware of Linux raising EIO when the slave is closed
+        child_output = bytearray()
+        while True:
+            try:
+                chunk = os.read(fd, 3000)
+            except OSError:  # Assume EIO
+                break
+            if not chunk:
+                break
+            child_output.extend(chunk)
+        os.close(fd)
+        return child_output.decode("ascii", "backslashreplace")
+
     def check_input_tty(self, prompt, terminal_input, stdio_encoding=None):
-        if not sys.stdin.isatty() or not sys.stdout.isatty():
-            self.skipTest("stdin and stdout must be ttys")
-        def child(wpipe):
-            # Check the error handlers are accounted for
-            if stdio_encoding:
-                sys.stdin = io.TextIOWrapper(sys.stdin.detach(),
-                                             encoding=stdio_encoding,
-                                             errors='surrogateescape')
-                sys.stdout = io.TextIOWrapper(sys.stdout.detach(),
-                                              encoding=stdio_encoding,
-                                              errors='replace')
-            print("tty =", sys.stdin.isatty() and sys.stdout.isatty(), file=wpipe)
-            print(ascii(input(prompt)), file=wpipe)
-        lines = self.run_child(child, terminal_input + b"\r\n")
-        # Check we did exercise the GNU readline path
-        self.assertIn(lines[0], {'tty = True', 'tty = False'})
-        if lines[0] != 'tty = True':
-            self.skipTest("standard IO in should have been a tty")
-        input_result = eval(lines[1])   # ascii() -> eval() roundtrip
+        template = (
+            'import sys, io\n'
+            '# Check the error handlers are accounted for\n'
+            'stdio_encoding = {stdio_encoding!a}\n'
+            'if stdio_encoding:\n'
+            '    sys.stdin = io.TextIOWrapper(sys.stdin.detach(),\n'
+            '                                 encoding=stdio_encoding,\n'
+            "                                 errors='surrogateescape')\n"
+            '    sys.stdout = io.TextIOWrapper(sys.stdout.detach(),\n'
+            '                                  encoding=stdio_encoding,\n'
+            "                                  errors='replace')\n"
+            '# Check we exercise the PyOS_Readline() path\n'
+            'if not sys.stdin.isatty() or not sys.stdout.isatty():\n'
+            '    raise AssertionError("standard IO should be a tty")\n'
+            'print("{sentinel}", end="")\n'
```

Comment:
It is expected that input() flushes stdout, not sure why you think this would need to be tested here. Avoiding the ``\n`` makes it clear that the length of ``sentinel`` on input is not tied to platform new line conversions.